### PR TITLE
feat(utils): add getFrameContexts method

### PR DESCRIPTION
**Pull request by WilcoFiers**
Add a utility which is passed a context object, and returns the context for each iframe.

Closes issue: #2937

### CODE REVIEW

File: test/core/utils/get-frame-context.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+      include: [['main']],
+      exclude: []
+    })
+    assert.deepEqual(contexts[2], {
+      include: [],
+      exclude: [['footer']]
+    })
+  });
+  
+  it.skip('mixes contexts if the frame is selected twice', function () {
```
Comment (WilcoFiers) [inline test/core/utils/get-frame-context.js:81]: There's a bug the presents this from passing. Need to fix that, but maybe in a separate PR.

File: lib/core/utils/get-frame-contexts.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import Context from '../base/context'
+import getAncestry from './get-ancestry'
+
+export default function getFrameContexts(context) {
+  const { frames } = new Context(context);
```
Comment (WilcoFiers) [inline lib/core/utils/get-frame-contexts.js:5]: This creates `axe._tree`, which is going to slow stuff down. Can we avoid that? And if not, how do we avoid either a tree that is outdated by the time axe.runPartial is called, or generating the tree twice.

Comment (straker) [reply lib/core/utils/get-frame-contexts.js:5]: Do we want to fix this in this pr or another one?

File: lib/core/base/context.js [hunk 3]
```diff
@@ -42,29 +42,51 @@ function pushUniqueFrame(collection, frame) {
...
+  //     include: [],
+  //     exclude: []
+  //   };
+
+  //   if (selectorArray) {
+  //     result[type].push(selectorArray);
+  //   }
+
+  //   context.frames.push(result);
+  // }
```
Comment (WilcoFiers) [inline lib/core/base/context.js:89]: I assume you didn't mean to leave that in.
```suggestion
```

File: lib/core/public/load.js [hunk 4]
```diff
@@ -13,6 +13,7 @@ function runCommand(data, keepalive, callback) {
   };
 
   var context = (data && data.context) || {};
+  console.log(data.context);
```
Comment (WilcoFiers) [inline lib/core/public/load.js:16]:
```suggestion
```

File: lib/core/utils/get-frame-contexts.js (hunk 2 above)
Comment (WilcoFiers) [reply lib/core/utils/get-frame-contexts.js:5]: Nope, we can let Context build axe._tree. I've accounted for it in the updated proposal.

File: test/core/utils/get-frame-context.js [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+  it('skips frames excluded by a parent', function() {
+    fixture.innerHTML = '<iframe></iframe>';
+    var frameContexts = getFrameContexts({
+      exclude: [['#fixture']]
+    });
+    assert.lengthOf(frameContexts, 0);
+  });
+
+  it('normalizes the context', function() {
```
Comment (WilcoFiers) [inline test/core/utils/get-frame-context.js:164]: Can you check the selector of these frames, make sure we get the right one?

File: test/core/utils/get-frame-context.js [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    });
+    assert.lengthOf(contexts, 1);
+    assert.deepEqual(contexts[0], {
+      initiator: false,
+      include: [],
+      exclude: []
+    });
+  });
+
+  it('accepts elements', function() {
```
Comment (WilcoFiers) [inline test/core/utils/get-frame-context.js:199]: Same here, check the selector.

File: test/testframe1.html [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<!DOCTYPE html>
```
Comment (WilcoFiers) [inline test/testframe1.html:1]: Delete please.

**Review by WilcoFiers — COMMENTED** [review]
LGTM

**Review by michael-siek — APPROVED** [review]

**Comment by WilcoFiers** [comment]
For the record: I passed this one off to Steve, who finished it up. Steve checked my work, and I checked his additions. Michael reviewed this for security.

**Pull request merged**